### ISSUE TITLE
TRGP: Trust Region Gradient Projection for Continual Learning

### ISSUE THREAD

**Issue by reyllama** [reyllama/paper-reviews#157]
## TL; DR

- Complement orthogonal gradient projection based CL method (i.e., Gradient Projection Memory) with incorporation of diagonal scaling matrix that adapts weights from most related previous task (=trust region).

## Motivation

![image](https://user-images.githubusercontent.com/46308407/153099651-7de0d9ef-db56-4d1a-9ee0-dc5fc48dfc86.png)

- Orthogonal gradient projection method works well only when tasks are un- or weakly correlated. Little further learning for task t occurs otherwise.

## Prior Works

- Gradient Projection Memory (ICLR'21)

### Trust Region

- For task t, find one or two previous tasks that have the highest correlation with the current task.
![image](https://user-images.githubusercontent.com/46308407/153099707-612528a4-a6de-41d7-af5c-64f42c2a226d.png)
![image](https://user-images.githubusercontent.com/46308407/153099720-90bc6d36-3fa9-4020-81a2-8cc3036273aa.png)

- High cosine similarity of gradient tensor w.r.t. previous task repr. subspace indicates high correlation between two tasks.

### Scaled Weight Projection

- Once we have trust regions for each task, we efficiently employ knowledge from the *related task* via diagonal scaling matrix (saved separately)
- Key intuition: Decompose weight gradient into two orthogonal components (one that is orthogonal to the trust region and one that belongs to the trust region repr. subspace)
- With separate scaling matrix Q_i for each task i, we can gain advantage for forward transfer as we leverage knowledge from relevant task.

![image](https://user-images.githubusercontent.com/46308407/153100151-67e5c187-4f4f-4f33-8592-d8335e887f8f.png)

### Task Subspace Construction

- Keep features from each layer for each task, and apply SVD and k-rank matrix approximation (sum of k rank-1 matrices).

### Continual Learning with TRGP

![image](https://user-images.githubusercontent.com/46308407/153100124-e41ea9ae-23dd-4b17-bf11-c34efa33f487.png)

## Experiments

- Evaluation protocol largely follows that of GPM (Saha et. al., ICLR'21)

![image](https://user-images.githubusercontent.com/46308407/153100225-93390f21-adaa-46c8-82d2-a6e1070fc400.png)

![image](https://user-images.githubusercontent.com/46308407/153100246-8c42d42a-4f07-4959-9d18-c41487c2ff76.png)
![image](https://user-images.githubusercontent.com/46308407/153100274-4d110e80-30a0-4c0c-8e5a-de06824b2398.png)
![image](https://user-images.githubusercontent.com/46308407/153100292-478fe015-e9a1-4dd6-a7d4-56227f2742d7.png)
![image](https://user-images.githubusercontent.com/46308407/153100303-67f4173b-18ae-4498-81db-e1f2726c307d.png)
![image](https://user-images.githubusercontent.com/46308407/153100317-df204114-7b55-40a3-a19d-411532ec9b38.png)
![image](https://user-images.githubusercontent.com/46308407/153100333-adc658e8-a003-4da0-b638-907b8c795fc7.png)
![image](https://user-images.githubusercontent.com/46308407/153100445-c3199764-3dbe-47ec-ae2e-e2733bab2b5b.png)

## Comments

- Clear and clever motivation (with two toy examples)
- Intuitive yet efficient solution proposed
- Satisfactory performance (at the cost of compute and memory)
